File: algorithm/LeetCode/Combination Sum/Combination Sum.java
public class Solution {
    List<List<Integer>> ret = new ArrayList<List<Integer>>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        for(int i = 0; i < candidates.length; ++i){
            ArrayList<Integer> tmpCdd = new ArrayList<Integer>();
            tmpCdd.add(candidates[i]);
            dfs(i, tmpCdd, candidates, target - candidates[i]);
            tmpCdd.remove(tmpCdd.size() - 1);
        }
        return ret;
    }

    public void dfs(int index, List<Integer> tmpCdd, int[] candidates, int target){
        if(index == candidates.length)
            return;
        if(target < 0)
            return; //直接剪枝
        if(target == 0){
            ret.add(new ArrayList<Integer>(tmpCdd));
            return;
        }else{
            for(int i = index; i < candidates.length; ++i){
                tmpCdd.add(candidates[i]);
                dfs(i, tmpCdd, candidates, target - candidates[i]);
                tmpCdd.remove(tmpCdd.size() - 1);
            }
        }
    }
}

File: algorithm/LeetCode/Combination Sum/Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        tmpCdd = candidates;
        sort(tmpCdd.begin(), tmpCdd.end());
        this->target = target;
        vector<int> tmpVec;
        dfs(0, tmpVec);
        return result;
    }    
private:
    int target;
    vector<int> tmpCdd;
    vector<vector<int>> result;
private:
    void dfs(int index, vector<int> & tmpVec)
    {
        if(index == tmpCdd.size()) return;    //到达叶节点
        int tmpSum = accumulate(tmpVec.begin(), tmpVec.end(), 0);
        if(tmpSum == target){
            result.push_back(tmpVec);
            return;
        }else if(tmpSum > target){//剪枝
            return;
        }else{
            for(int i = index; i < tmpCdd.size(); ++i){//这里从i开始的原因是因为参数可以是重复的
                tmpVec.push_back(tmpCdd[i]);
                dfs(i, tmpVec);
                tmpVec.pop_back();//回溯
            }
        }
    }
};